### DISCUSSION TITLE
Halfedge data structure: questions about the order of halfedges

### DISCUSSION THREAD

**Discussion by cjamin** [CGAL/cgal#8458] [Q&A] [answered]
Hi,

I have two questions regarding the HDS:

1. Can I assume that in the list of halfedges returned by `halfedge_handles()`, each even-indexed halfedge is followed by its opposite halfedge? I saw in the [assignement operator implementation](https://github.com/CGAL/cgal/blob/eac5679444a9056fda520a47a130119f00032e2b/HalfedgeDS/include/CGAL/HalfedgeDS_list.h#L420) that it seems to be the case, but is it documented somewhere?
2. When copying a HDS (using the assignement operator mentioned above, or the copy constructor), is the order of the halfedges preserved? If yes (the code seems to say it is), is it guaranteed in all cases? It could be nice to document it. In my case, I have a list of object that is stored separately. Each object stores a halfedge handle pointing to the halfedge it is attached to. I need to backup the HDS and this list, and I want a efficient way to restore the link between the objects and the HDS.

Thanks!

**Comment by lrineau** [comment, marked as answer]
Hi Clément, it seems those properties are indeed ensured by the current code, but not documented.

That means the authors of the code are allowed to change that behavior in future versions of CGAL.

That kind of non-documented invariant are quite painful, because the user code is usually more performant if it can assume it.

**Comment by afabri** [comment]
It's the same/simliar in `Surface_mesh` but not in `Linear_cell_complex`.

**Comment by cjamin** [comment]
Thank you both for your answers.
What would it take to document it?
